file-format.md

```diff
@@ -0,0 +1,59 @@
+# timesheet.txt File Format Specification
+
+**Please note that this specification is preliminary and a work in progress.**
+It will be finalized in [the 1.0 release](https://github.com/scy/timesheet.txt/milestone/4).
+Therefore, the wording might be terse in some parts.
+
+**But!**
+If you think anything is confusing, unclear or even wrong, please open an [issue](https://github.com/scy/timesheet.txt/issues) so that we can clarify it before release.
+
+## Purpose
+
+_timesheet.txt_ defines a plain text file format for time sheets, i.e. a log of what a single person did during specific intervals in time.
+
+Please refer to the [readme](README.md) for more information about its design goals and philosophy, as they will not be repeated here.
+
+## Terms
+
+To make sure that we're talking about the same things, this specification will define some special terms that have a certain meaning when it comes to our file format.
+These terms will be written in ***bold italic*** font to make them clearly distinguishable.
+
+Unfortunately, it's a bit easy to mix up these terms in a colloquial setting.
+Also, the format is still in the process of being specified, therefore it can happen that the wrong term is used in a GitHub issue or even project documentation.
+You're welcome to point these out for us to fix.
+
+## The Input File
+
+The _timesheet.txt_ tools act upon a stream of data called the ***input file***.
+This does not have to be a real file, since the tools are designed to be able to read from a stream (e.g. Unix's *standard input*), too.
+Since we don't care how you create this stream of data, and it's transparent to the tools, it will nevertheless be called the input file.
+
+The characters in the input file are expected to be encoded according to the UTF-8 standard.
+
+The input file is read from start to end, one line at a time.
+A ***line*** ends with a ***line terminator***, a newline character (`\n` or `LF` or `U+000A`).
+Since whitespace is removed from the end of each line (see below), Windows-style line endings (`\r\n` or `CRLF` or `U+000D` followed by `U+000A`) are implicitly supported as well.
+Mac OS 9 style line endings (`CR` only) are not, though.
+
+It is a good habit to end the file with a line terminator, but this is not required.
```

Comment:
If a file does not need to end with a line terminator, it means that lines (as defined by the previous paragraph) do not always end in a `\n`. The last line of the file will end at EOF instead.

Thinking about parsers (and writers alike) having to be simple, is there any reason not to put a strict requirement on the ending `\n` of a file? It would definitely make appending new lines a whole lot easier.

Reply:
> The last line of the file will end at EOF instead.

Not "will" but "can".

> is there any reason not to put a strict requirement on the ending `\n` of a file?

Yes: Some editors don't do it by default, and while I agree that it's good Unix convention to always end a file with a newline, I'm not sure I want to force this requirement on users who maybe don't have a Unix background at all or simply forgot to configure their editor in a certain way.

I would say that it doesn't complicate a parser significantly to accept a non-newline-terminated line at the end of the file. In fact I'm pretty sure most higher level languages will not require a terminating newline in their builtin "read file line by line" implementations.

But I agree that it makes a _writer_ more complicated, because before appending a line, it has to check whether the file ends in a newline and, if not, append one. It's not horrendously complicated though: Open the file for reading and writing (instead of just for writing), seek to length minus one, read the byte, if it's not `\n`, add it.

Also, this will only be a problem if the file has been edited with _both_ a tool (or editor) that doesn't enforce the final newline _and_ a tool that thinks there should be one. In other words, if you always use a writer tool that ends each line with `\n`, you won't have any problems either.

TL;DR: I don't think that adding ~3 lines of code to a writer justifies eating the last line for people with poorly configured editors.

I'd nevertheless add a warning to my Python implementation that will be printed if the last line doesn't end in `\n` and suggest in the specification that other implementations should do the same. Is that a compromise you could live with?

Reply:
I think I am fine with all outcomes. Mostly nitpicking at spec language here as I dislike ambiguities.

I guess what I would do is rewrite the definition of a line to make it clear it can end with either `\n` or EOF, and then (if you want to go with [RFC 2119](https://tools.ietf.org/html/rfc2119) terminology) note that ending a file SHOULD (recommendation, rather than a MUST obligation) end on a `\n`.

Reply:
I think I am fine with all outcomes. Mostly nitpicking at spec language here as I dislike ambiguities.

I guess what I would do is rewrite the definition of a line to make it clear it can end with either `\n` or EOF, and then (if you want to go with [RFC 2119](https://tools.ietf.org/html/rfc2119) terminology) note that ending a file SHOULD (recommendation, rather than a MUST obligation) end on a `\n`.